software/contrib/bouncing_pixels.md

```diff
@@ -0,0 +1,201 @@
+
+Denoted gate hold lengths are defaults. They can be changed in the
+[configuration file](#Configuration).
+
+## Controls
+
+* K1: simulation speed
+* K2: impulse speed
+* Either button + K1: ball count
+* Either button + K2: width
+* B1 short press: reset
+* B2 short press: send impulse
+* Digital input: send impulse (customisable)
+* Analogue input: speed (customisable)
+
+Second layer knobs are "locking", which is to say you need to physically set the
+knob to a position close to the registered value in order for it to latch and
+start changing. This is to prevent abrupt changes in values when you press the
+buttons. If you lose track of a knob position, try sweeping the full range to
+get it to latch.
+
+The short press threshold is set to 500ms by default. You can adjust this
+length in the [configuration file](#Configuration).
+
+The digital input can be set to trigger a reset instead of an impulse. Likewise,
+the analogue input can be set to control any of the parameters controlled by
+knobs. These behaviours can be defined [configuration file](#Configuration).
+
+## Configuration
+
+The script can be thoroughly customised using a configuration file. The
+configuration file must be located at `/config/BouncingPixels.json` on the micro
+controller.
+
+### Sample configuration file
+
+```json
+{
+    "long_press_length": 500.0,
+    "timescale_min": 0.0,
+    "timescale_max": 100.0,
+    "knob_change_threshold": 0.01,
+    "din_function": "impulse",
+    "ain_function": "speed",
+    "gate_hold_length_top": 25.0,
+    "gate_hold_length_left": 25.0,
+    "gate_hold_length_right": 25.0,
+    "gate_hold_length_bottom": 25.0,
+    "gate_hold_length_any": 10.0,
+    "gate_hold_length_corner": 100.0,
+    "arena_width_min": 30.0,
+    "arena_width_max": 1920.0,
+    "ball_count_max": 100,
+    "ball_count_min": 1,
+    "corner_collision_margin": 15.0,
+    "start_speed_min": 10.0,
+    "start_speed_max": 100.0,
+    "accel_min": -5.0,
+    "accel_max": 5.0,
+    "bounciness_min": 0.8,
+    "bounciness_max": 1.2,
+    "bounce_angle_deviation_max": 15.0,
+    "under_speed_behaviour": "reset",
+    "over_speed_behaviour": "reset",
+    "under_speed_threshold": 5.0,
+    "over_speed_threshold": 1000000,
+    "impulse_speed_min": 0,
+    "impulse_speed_max": 100000,
+    "impulse_speed_variation_min": 0.5,
+    "impulse_speed_variation_max": 2.0,
+}
+```
+
+### Configuration values
+
+| Key                           | Type                  | Possible values                                        | Default value | Description |
+|-------------------------------|-----------------------|--------------------------------------------------------|---------------|-------------|
+| `poll_frequency`              | Floating point number | 5 - 120                                                | 20            | How frequently the application polls for new inputs, expressed as times per second.<br><br>⚠️ **Changing this value is not recommended.** |
```

Comment:
Why limit the polling frequency, instead of just polling the inputs as fast as possible so you can respond immediately? I haven't read through the code exhaustively, so maybe that's explained elsewhere and I just haven't read that yet.